sql: move external span construction into span_builder

michae2 opened the pull request:
**sql: move external span construction into span_builder**

In #126875 we started rewriting spans to external spans in the
kv_batch_fetcher and streamer when querying tables with external row
data. This span rewrite is very late, after the DistSQL physical planner
has already chosen where to plan DistSQL flows by consulting the range
cache.

This commit moves external span construction earlier, from the
kv_batch_fetcher and streamer into the span_builder. The span_builder is
used both at execbuild time for scans and at execution time for lookup
and zigzag joins.

By moving external span construction earlier, before DistSQL physical
planning, we allow the physical planner to work with spans in the same
form as the range cache. This helps the DistSQL physical planner make
the correct decisions about where to plan flows.

This commit also happens to fix a rare bug when the same spans were
rewritten multiple times in the kv_batch_fetcher, by deleting that code.

Fixes: #128070

Release note: None

---

**sql: add TestExternalRowDataDistSQL**

Add a test that the DistSQL physical planner correctly plans flows
reading from external table data.

Release note: None

michae2 committed:
sql: move external span construction into span_builder

In #126875 we started rewriting spans to external spans in the
kv_batch_fetcher and streamer when querying tables with external row
data. This span rewrite is very late, after the DistSQL physical planner
has already chosen where to plan DistSQL flows by consulting the range
cache.

This commit moves external span construction earlier, from the
kv_batch_fetcher and streamer into the span_builder. The span_builder is
used both at execbuild time for scans and at execution time for lookup
and zigzag joins.

By moving external span construction earlier, before DistSQL physical
planning, we allow the physical planner to work with spans in the same
form as the range cache. This helps the DistSQL physical planner make
the correct decisions about where to plan flows.

This commit also happens to fix a rare bug when the same spans were
rewritten multiple times in the kv_batch_fetcher, by deleting that code.

Fixes: #128070

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132139)
<!-- Reviewable:end -->

michae2 committed:
sql: add TestExternalRowDataDistSQL

Add a test that the DistSQL physical planner correctly plans flows
reading from external table data.

Release note: None

DrewKimball reviewed (approved):
Overall <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - just some questions.

> Reviewed 13 of 13 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-:-O8nlnds7bgx5Q2NEzkC:byc705k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @michae2, and @msbutler)

___
*[`pkg/sql/catalog/tabledesc/structured.go` line 2213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-O8nh0Ru5gSp3PAwtGMe:-O8nh0Ru5gSp3PAwtGMf:brwr21d) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ff3fbd8a297d0ba5f8c17dbc83a909e76e1dc6b/pkg/sql/catalog/tabledesc/structured.go#L2213)):*
> ```Go
> // IndexSpan implements the TableDescriptor interface.
> func (desc *wrapper) IndexSpan(codec keys.SQLCodec, indexID descpb.IndexID) roachpb.Span {
> 	prefix := roachpb.Key(rowenc.MakeIndexKeyPrefix(codec, desc.GetID(), indexID))
> ```

Could we / should we assert that `ExternalRowData` is nil for these variants? How bad is it if we accidentally don't call the `AllowingExternalRowData` version in a place where we should?
___
*[`pkg/sql/span/span_builder.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-O8nitUi2Af5MT9XaE6O:-O8nitUi2Af5MT9XaE6P:blspugx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ff3fbd8a297d0ba5f8c17dbc83a909e76e1dc6b/pkg/sql/span/span_builder.go#L44)):*
> ```Go
> 
> // Init initializes a Builder with a table and index. It does not set up the
> // Builder to create external spans, even if the table uses external row data.
> ```

Do we know under what circumstances we'd want to build spans without using the external prefix?
___
*[`pkg/sql/row/external_row_data_test.go` line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-O8nlh5L4brzsJ1DwG-4:-O8nlh5MEuOD-YnNUtu-:bb4ngfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c95a676b710702e243f9a58b083497214619ef00/pkg/sql/row/external_row_data_test.go#L117)):*
> ```Go
> func TestExternalRowDataDistSQL(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	defer log.Scope(t).Close(t)
> ```

Should we preemptively disable this test under duress?


<!-- Sent from Reviewable.io -->

dt commented on pkg/sql/catalog/fetchpb/index_fetch.proto:
```diff
@@ -130,13 +131,13 @@ message IndexFetchSpec {
-    // keys that need to be used against the KV layer to fetch the external
-    // data.
-    optional bytes new_prefix = 3;
+    // TenantID is the new tenant ID to use in keys when fetching external data
```
actually, you know, this will just error out with wrong wire type if we do this in place which is maybe nice (and loud) compared to the alternative which would be a gate. Carry on.

michae2 reviewed (commented):
Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-:-O8rqRKAD-CF8YGNwwJo:b-ucfknh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @dt, and @msbutler)

___
*[`pkg/sql/catalog/fetchpb/index_fetch.proto` line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/132139#gh-1795613322:-O8ri2PuCD_Zz-jarx71:b-gt8dmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c95a676b710702e243f9a58b083497214619ef00/pkg/sql/catalog/fetchpb/index_fetch.proto#L134)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

actually, you know, this will just error out with wrong wire type if we do this in place which is maybe nice (and loud) compared to the alternative which would be a gate. Carry on.
</blockquote></details>

Thanks! I'm happy to avoid the gate :slightly_smiling_face:
___
*[`pkg/sql/catalog/tabledesc/structured.go` line 2213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-O8nh0Ru5gSp3PAwtGMe:-O8riBQNBJkOeUh1fx6p:b4shcch) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ff3fbd8a297d0ba5f8c17dbc83a909e76e1dc6b/pkg/sql/catalog/tabledesc/structured.go#L2213)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could we / should we assert that `ExternalRowData` is nil for these variants? How bad is it if we accidentally don't call the `AllowingExternalRowData` version in a place where we should?
</blockquote></details>

Great idea, I've added some assertions.
___
*[`pkg/sql/span/span_builder.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-O8nitUi2Af5MT9XaE6O:-O8riJE2A7GafSTP1fCf:b-hok6t5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ff3fbd8a297d0ba5f8c17dbc83a909e76e1dc6b/pkg/sql/span/span_builder.go#L44)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Do we know under what circumstances we'd want to build spans without using the external prefix?
</blockquote></details>

So far it's only for DeleteRange, every other use is for what will be a ScanRequest or GetRequest: https://github.com/cockroachdb/cockroach/blob/81a1dfe888b33158e8a282b4ced2bf6e4eddc1ac/pkg/sql/opt_exec_factory.go#L1788

I added an assertion that the table doesn't use external row data.
___
*[`pkg/sql/row/external_row_data_test.go` line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-O8nlh5L4brzsJ1DwG-4:-O8rqO2fE_oA7Rz3kVVM:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c95a676b710702e243f9a58b083497214619ef00/pkg/sql/row/external_row_data_test.go#L117)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we preemptively disable this test under duress?
</blockquote></details>

Good call, done.


<!-- Sent from Reviewable.io -->

michae2 commented:
bors r=DrewKimball,dt

yuzefovich reviewed (commented):
Thanks for cleaning this up after me :smiley: 

> Reviewed 10 of 13 files at r1, 6 of 6 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132139#-:-OAzAyrlBbrBinpt927k:b-wf25zm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->
